### ISSUE TITLE
trame server launch and program input arguments

### ISSUE THREAD

**Issue by francesco-ballarin** [pyvista/pyvista#3973]
I have the following `test_one.py` file

```
import pyvista
pyvista.set_jupyter_backend("client")

def test_one():
  pass
```

which I plan to run with `pytest`.

As soon as I add options to `pytest`, e.g. `pytest -svv` I get the following:

```
==================================================== test session starts =====================================================
platform linux -- Python 3.10.9, pytest-7.1.2, pluggy-1.0.0+repack -- /usr/bin/python3
cachedir: .pytest_cache
benchmark: 3.2.2 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /tmp/test
plugins: nbval-0.9.6, cov-3.0.0, anyio-3.6.1, instafail-0.4.2, xdist-3.1.0, dependency-0.5.1, benchmark-3.2.2
collecting ... usage: pytest [-h] [--server] [--app] [--no-http] [--authKeyFile AUTHKEYFILE] [--hot-reload] [-d] [-s] [-i HOST] [-p PORT]
              [-t TIMEOUT] [-c CONTENT] [-a AUTHKEY] [-ws WS] [--no-ws-endpoint] [--fs-endpoints FSENDPOINTS]
              [--reverse-url REVERSE_URL] [--ssl SSL]
pytest: error: argument -s/--nosignalhandlers: ignored explicit argument 'vv'
collected 0 items / 1 error                                                                                                  

=========================================================== ERRORS ===========================================================
_________________________________________________ ERROR collecting test_1.py _________________________________________________
/usr/lib/python3.10/argparse.py:1871: in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
/usr/lib/python3.10/argparse.py:2084: in _parse_known_args
    start_index = consume_optional(start_index)
/usr/lib/python3.10/argparse.py:1992: in consume_optional
    raise ArgumentError(action, msg % explicit_arg)
E   argparse.ArgumentError: argument -s/--nosignalhandlers: ignored explicit argument 'vv'

During handling of the above exception, another exception occurred:
test_1.py:2: in <module>
    pyvista.set_jupyter_backend("client")
/usr/local/lib/python3.10/dist-packages/pyvista/jupyter/__init__.py:211: in set_jupyter_backend
    elegantly_launch(pyvista.global_theme.trame.jupyter_server_name)
/usr/local/lib/python3.10/dist-packages/pyvista/trame/jupyter.py:373: in elegantly_launch
    return asyncio.run(launch_it())
/usr/local/lib/python3.10/dist-packages/nest_asyncio.py:35: in run
    return loop.run_until_complete(task)
/usr/local/lib/python3.10/dist-packages/nest_asyncio.py:83: in run_until_complete
    self._run_once()
/usr/local/lib/python3.10/dist-packages/nest_asyncio.py:119: in _run_once
    handle._run()
/usr/lib/python3.10/asyncio/events.py:80: in _run
    self._context.run(self._callback, *self._args)
/usr/local/lib/python3.10/dist-packages/nest_asyncio.py:195: in step
    step_orig(task, exc)
/usr/lib/python3.10/asyncio/tasks.py:232: in __step
    result = coro.send(None)
/usr/local/lib/python3.10/dist-packages/pyvista/trame/jupyter.py:368: in launch_it
    await launch_server(server)
/usr/local/lib/python3.10/dist-packages/pyvista/trame/jupyter.py:133: in launch_server
    server.start(
/usr/local/lib/python3.10/dist-packages/trame_server/core.py:444: in start
    options = self.cli.parse_known_args()[0]
/usr/lib/python3.10/argparse.py:1874: in parse_known_args
    self.error(str(err))
/usr/lib/python3.10/argparse.py:2599: in error
    self.exit(2, _('%(prog)s: error: %(message)s\n') % args)
/usr/lib/python3.10/argparse.py:2586: in exit
    _sys.exit(status)
E   SystemExit: 2
```

It seems to me that arguments that I pass to `pytest` are somehow also passed to the trame server launcher.
Should I somehow guard the call to `pyvista.set_jupyter_backend`, since obviously while running with `pytest` the goal is not to produce jupyter widgets? I don't seem to remember having to do that for older backends, and if possible I would not like to guard that call.

Thanks!

**Comment by banesullivan** [comment]
This looks like a `trame_server` issue with it parsing the invocation arguments on launch. @jourdain, any tips on how to avoid this?

**Comment by jourdain** [comment]
The question is why it complains. Normally it should not `crash/exit` even if other args are provided.

The -s from wslink conflict with `-svv`. We do want to parse any args but we should make sure we prevent conflicts and exception/error. 

I'm not sure how to properly sanitize those args with argparse (either in wslink or trame-server).

**Comment by banesullivan** [comment]
In the meantime, a quick fix would be to put the backend launch in a pytest fixture or some sort of other guard. Our tests simply set the backend in each test:

https://github.com/pyvista/pyvista/blob/bdb60e1f7ff8872ba9073d194825aa9c1726ced7/tests/jupyter/test_trame.py#L44-L45

Perhaps this fixture would work for you

```py
import pyvista

@fixture()
def trame_backend():
    pyvista.set_jupyter_backend("client")


def test_one(trame_backend):
    ...  # do stuff
```
```

**Comment by francesco-ballarin** [comment]
I don't think that will make much difference:

```
import pyvista as pv
import pytest

def test_one():
    pl = pv.Plotter(notebook=True)
    pl.add_mesh(pv.Cone())
    pl.show(jupyter_backend="client")

@pytest.fixture()
def trame_backend():
    pv.set_jupyter_backend("client")

def test_two(trame_backend):
    pl = pv.Plotter(notebook=True)
    pl.add_mesh(pv.Cone())
    pl.show()
```

I'm getting the same error as in the first post, twice.

**Comment by jourdain** [comment]
`trame/server/client` backend will trigger that error.

**Comment by banesullivan** [comment]
Oh. Huh. I just realized that I am unable to reproduce this at all. Would you please share a `pv.Report()` so I can see all your package versions:

```py
import pyvista as pv
print(pv.Report())
```

**Comment by jourdain** [comment]
You can test/validate by running any trame script with the `-svv` arg...

**Comment by banesullivan** [comment]
@jourdain, I find it a bit odd that the `Server` class would have any awareness of the command-line arguments and always parse them within `Server.start`: https://github.com/Kitware/trame-server/blob/315cb9b1e0bcb33d7944f963ae8470495e59be6c/trame_server/core.py#L479

Perhaps the CLI could be a separate interface in `trame-server` that would then construct the `Server` object with the provided options.

Otherwise, I cannot pin down why the argument parser is failing in this case... I think there is something being set by `pytest` that causes this incompatibility as using trame in other invocations with the `-svv` options works fine

**Comment by jourdain** [comment]
https://github.com/Kitware/wslink/blob/master/python/src/wslink/server.py#L40-L45

**Comment by psavery** [comment]
`pytest` won't allow any extra arguments to be used with it (if you try to add any `trame` arguments, there will be an error), so maybe we should just disable argument parsing in `trame` if it is being ran with `pytest`?

As per some discussion with @banesullivan, we were thinking to have a `TRAME_NO_ARG_PARSE` environment variable that skips argument parsing (although the relevant settings will still be present on the server, but they will be their defaults). We can also check if we are running `pytest` and default this environment variable to `True` if that is the case.

Any thoughts?

**Comment by francesco-ballarin** [comment]
> We can also check if we are running pytest and default this environment variable to True if that is the case.

Looks good to me!

**Comment by psavery** [comment]
This is fixed in kitware/trame-server#14, although it is a little different from what was mentioned above.

The arguments are still ignored if `pytest` is being used, so that should fix this issue. However, instead of having a `TRAME_NO_ARG_PARSE` environment variable, we now have a `TRAME_ARGS` environment variable where you can specify the arguments for trame (such as `TRAME_ARGS="-p 8080 --server"`). This would allow someone to specify trame arguments to be used even in a pytest environment. Or even `TRAME_ARGS=""` can be used explicitly so that trame won't parse any arguments. There is also an argument version of this as well: `--trame-args="-p 8080 --server"`.
